semantic_search/main.py

```diff
@@ -71,24 +80,60 @@ def app_startup():
     model.tokenizer, model.model = setup_model_and_tokenizer(
         settings.pretrained_model_name_or_path, cuda_device=settings.cuda_device
     )
-    model.similarity = torch.nn.CosineSimilarity(-1)
+    embedding_dim = model.model.config.hidden_size
+    model.index = setup_faiss_index(embedding_dim)
 
 
 @app.post("/")
 async def query(query: Query) -> List[Dict[str, float]]:
-    ids = [query.query.uid] + [document.uid for document in query.documents]
     text = [query.query.text] + [document.text for document in query.documents]
 
     embeddings = encode(text)
-    similarity_scores = model.similarity(embeddings[0], embeddings[1:])
 
-    # If top_k not specified, return all documents.
-    top_k = similarity_scores.size(0)
+    query_id, query_embedding = query.query.uid, embeddings[0]
+    document_ids, document_embeddings = [doc.uid for doc in query.documents], embeddings[1:]
+
+    query_embedding = np.asarray(query_embedding).reshape(1, -1).astype("float32")
+    document_embeddings = (
+        np.asarray(document_embeddings).reshape(-1, model.index.d).astype("float32")
+    )
+
+    query_id = (
+        np.asarray(query_id)
+        .reshape(
+            -1,
+        )
+        .astype("int64")
+    )
+    document_ids = (
+        np.asarray(document_ids)
+        .reshape(
+            -1,
+        )
+        .astype("int64")
+    )
+
+    # Only add items to the index if they do not already exist.
+    # See: https://github.com/facebookresearch/faiss/issues/859
+    for doc_id, doc_embedding in zip(document_ids, document_embeddings):
+        if doc_id not in faiss.vector_to_array(model.index.id_map):
+            doc_embedding = doc_embedding.reshape(1, -1)
+            doc_id = doc_id.reshape(  # type: ignore
+                1,
+            )
+            model.index.add_with_ids(doc_embedding, doc_id)
```

Comment:
I don't love that we add ids one by one. Should probably batch this.

Reply:
I have updated this so that embeddings are added to FAISS in a batch. Shaved about 200ms of each request. I think this code should be moved to its own function though, maybe in `common.util`.